src/tools/chip-cert/Cmd_PrintCD.cpp

```diff
@@ -0,0 +1,277 @@
+    }
+
+    return kCDFormat_Unknown;
+}
+
+bool ReadCD(const char * fileNameOrStr, MutableByteSpan cd)
+{
+    CDFormat cdFmt = kCDFormat_Unknown;
+    uint32_t cdLen = 0;
+    std::unique_ptr<uint8_t[]> cdBuf;
+
+    // If fileNameOrStr is a file name
+    if (access(fileNameOrStr, R_OK) == 0)
+    {
+        VerifyOrReturnError(ReadFileIntoMem(fileNameOrStr, nullptr, cdLen), false);
+
+        cdBuf = std::unique_ptr<uint8_t[]>(new uint8_t[cdLen]);
+
+        VerifyOrReturnError(ReadFileIntoMem(fileNameOrStr, cdBuf.get(), cdLen), false);
+
+        cdFmt = DetectCDFormat(cdBuf.get(), cdLen);
+        if (cdFmt == kCDFormat_Unknown)
+        {
+            fprintf(stderr, "Unrecognized CD Format in File: %s\n", fileNameOrStr);
+            return false;
+        }
+    }
+    // Otherwise, treat fileNameOrStr as a pointer to the certificate string
+    else
+    {
+        cdLen = static_cast<uint32_t>(strlen(fileNameOrStr));
+
+        cdFmt = DetectCDFormat(reinterpret_cast<const uint8_t *>(fileNameOrStr), cdLen);
+        if (cdFmt == kCDFormat_Unknown)
+        {
+            fprintf(stderr, "Unrecognized CD Format in the Input Argument: %s\n", fileNameOrStr);
+            return false;
+        }
+
+        cdBuf = std::unique_ptr<uint8_t[]>(new uint8_t[cdLen]);
+        memcpy(cdBuf.get(), fileNameOrStr, cdLen);
+    }
+
+    if (cdFmt == kCDFormat_Hex)
+    {
+        size_t len = chip::Encoding::HexToBytes(Uint8::to_char(cdBuf.get()), cdLen, cdBuf.get(), cdLen);
+        VerifyOrReturnError(CanCastTo<uint32_t>(2 * len), false);
+        VerifyOrReturnError(2 * len == cdLen, false);
+        cdLen = static_cast<uint32_t>(len);
+    }
+    else if (cdFmt == kCDFormat_Base64)
+    {
+        VerifyOrReturnError(Base64Decode(cdBuf.get(), cdLen, cdBuf.get(), cdLen, cdLen), false);
+    }
+
+    VerifyOrReturnError(cdLen <= cd.size(), false);
+    memcpy(cd.data(), cdBuf.get(), cdLen);
+
+    cd.reduce_size(cdLen);
+
+    return true;
+}
+
+void ENFORCE_FORMAT(1, 2) SimpleDumpWriter(const char * aFormat, ...)
+{
+    va_list args;
+
+    va_start(args, aFormat);
+
+    vprintf(aFormat, args);
+
+    va_end(args);
+}
+
+bool PrintCD(const char * fileName, ByteSpan cd)
+{
+    FILE * file = nullptr;
+    chip::TLV::TLVReader reader;
+    ByteSpan signerKeyId;
+    ByteSpan cdContent;
+    std::unique_ptr<char[]> signerKeyIdHex;
+    uint32_t signerKeyIdHexLen = 0;
+
+    VerifyOrReturnError(!cd.empty(), false);
+    VerifyOrReturnError(OpenFile(fileName, file, true), false);
+    VerifyOrReturnError(CMS_ExtractKeyId(cd, signerKeyId) == CHIP_NO_ERROR, false);
+    VerifyOrReturnError(CMS_ExtractCDContent(cd, cdContent) == CHIP_NO_ERROR, false);
+
+    signerKeyIdHexLen = 2 * static_cast<uint32_t>(signerKeyId.size()) + 1;
+    signerKeyIdHex    = std::unique_ptr<char[]>(new char[signerKeyIdHexLen]);
+    VerifyOrReturnError(BytesToHex(signerKeyId.data(), signerKeyId.size(), signerKeyIdHex.get(), signerKeyIdHexLen,
```

Comment:
`BytesToUppercaseHexString` avoids needing the HexFlags and is clearer up-front.

Reply:
updated